### ISSUE TITLE
Error: MPMD detected but reload is not supported yet for neuron distributed environment

### ISSUE THREAD

**Issue by wfckl789** [aws-neuron/neuronx-distributed#21]
Hi, I found the error `MPMD detected but reload is not supported yet` will occur if I open `Eager Debug Mode` for a model trained in neuron distributed environment where dp=1, tp=8, pp=4. Could you help check this issue? **Thanks so much!**

![image](https://github.com/aws-neuron/aws-neuron-sdk/assets/42508752/79763642-1ea9-4356-b4f8-4055fadd0de3)

I attach related scripts here and you can simply run `./run_simple_model_tp_pp.sh ` after download them.

[scripts.zip](https://github.com/aws-neuron/aws-neuron-sdk/files/15209167/scripts.zip)

Environment information:

EC2 Instance: trn1.32.xlarge

OS: Ubuntu 20.04

Neuron Pytorch: Latest 2.18

**Comment by aws-rhsoln** [comment]
Thank you for reporting the issue. `MPMD detected` error means that 1 worker is trying to load a new graph vs the other worker is waiting to perform inference, and hence the graphs that have collectives are not able to communicate with each other. Here the assumption is that each worker performs the same set of operations and hence all workers would be in SPMD mode. This assumption is broken and hence you see the above error. 
We are looking into the scripts now to identify which worker is producing a new graph (and why) and if we can modify the scripts such that graphs with collectives do not change from iteration to another. Will report back once we have an update.

**Comment by wfckl789** [comment]
> Thank you for reporting the issue. `MPMD detected` error means that 1 worker is trying to load a new graph vs the other worker is waiting to perform inference, and hence the graphs that have collectives are not able to communicate with each other. Here the assumption is that each worker performs the same set of operations and hence all workers would be in SPMD mode. This assumption is broken and hence you see the above error. We are looking into the scripts now to identify which worker is producing a new graph (and why) and if we can modify the scripts such that graphs with collectives do not change from iteration to another. Will report back once we have an update.

Thanks for your detailed explain! And I'd like to know that: Will this error interrupt the process of generating hlo graphs? 
I mean, my assumption is that I can't get all hlo graphs because the graph generation process will be shut down due to this error. Is my assumption right?

**Comment by aws-rhsoln** [comment]
So at a time, only one graph can be executed. In this case, since the run errored out at this graph, you can generate only graphs upto this point. If you want to generate all the graphs without worrying about execution, you can run with [neuron_parallel_compile](https://awsdocs-neuron.readthedocs-hosted.com/en/latest/frameworks/torch/torch-neuronx/api-reference-guide/training/pytorch-neuron-parallel-compile.html?highlight=neuron_parallel_compile#pytorch-neuronx-neuron-parallel-compile-cli) . The utility should help to extract all the HLOs, compile them and save them in the cache.

**Comment by jeffhataws** [comment]
@wfckl789 just want to check in to see if you were you able to make forward progress?

**Comment by wfckl789** [comment]
> @wfckl789 just want to check in to see if you were you able to make forward progress?

In this case as I obeserved, the cc compiler raised this compilation fault before the forward progress was executed, because I didn't see the loss value.

**Comment by aws-rhsoln** [comment]
Any particular reason for trying eager mode in mutli-worker case? Note: Eager debug mode is only for debugging and is not the most performant mode of execution
